### PR TITLE
Add more counters for file handle cache to understand it better

### PR DESCRIPTION
**Pull request by tanjialiang**

### CODE REVIEW

File: velox/connectors/hive/FileHandle.h [hunk 1]
```diff
@@ -67,14 +67,35 @@ using FileHandleCache = SimpleLRUCache<std::string, FileHandle>;
...
+    return {
+        minGenerationTimeMcrs_,
+        maxGenerationTimeMcrs_,
+        totalGenerationTimeMcrs_,
+        numGeneration_};
+  }
 
  private:
   const std::shared_ptr<const Config> properties_;
+  std::mutex m_;
```
Comment (spershin) [inline velox/connectors/hive/FileHandle.h:94]: If we use m_ only for stats, should we call it statsMutex_?

File: velox/common/caching/SimpleLRUCache.h [hunk 2]
```diff
@@ -36,125 +26,147 @@ namespace facebook::velox {
...
+  size_t numAccumulativeHit_{0};
+  size_t numAccumulativeTotal_{0};
 
   struct Element {
     Key key;
     Value* value;
-    int size;
-    int pinCount;
+    size_t size;
+    size_t pinCount;
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/common/caching/SimpleLRUCache.h:169]: Should be not much difference this is not a big structure.

File: velox/connectors/hive/FileHandle.cpp [hunk 3]
```diff
@@ -38,16 +39,33 @@ std::string groupName(const std::string& filename) {
...
-                         ->openFileForRead(filename);
-  fileHandle->uuid = StringIdLease(fileIds(), filename);
-  fileHandle->groupId = StringIdLease(fileIds(), groupName(filename));
-  VLOG(1) << "Generating file handle for: " << filename
-          << " uuid: " << fileHandle->uuid.id();
-  // TODO: build the hash map/etc per file type -- presumably after reading
-  // the appropriate magic number from the file, or perhaps we include the file
-  // type in the file handle key.
+  std::unique_ptr<FileHandle> fileHandle;
+  uint64_t timeElapsedMcrs;
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/connectors/hive/FileHandle.cpp]: Going to remove this. Will add extra counter exporting for the elapsed time.

File: velox/connectors/hive/FileHandle.cpp [hunk 4]
```diff
@@ -38,16 +39,33 @@ std::string groupName(const std::string& filename) {
...
+    // the appropriate magic number from the file, or perhaps we include the
+    // file type in the file handle key.
+  }
+  {
+    // Updating counters.
+    std::lock_guard<std::mutex> l(m_);
+    if (timeElapsedMcrs < minGenerationTimeMcrs_) {
+      minGenerationTimeMcrs_ = timeElapsedMcrs;
+    }
+    if (timeElapsedMcrs > maxGenerationTimeMcrs_) {
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/connectors/hive/FileHandle.cpp]: will be gone

File: velox/connectors/hive/FileHandle.cpp [hunk 5]
```diff
@@ -38,16 +39,33 @@ std::string groupName(const std::string& filename) {
...
+      maxGenerationTimeMcrs_ = timeElapsedMcrs;
+    }
+    totalGenerationTimeMcrs_ += timeElapsedMcrs;
+    ++numGeneration_;
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/connectors/hive/FileHandle.cpp]: will be gone

File: velox/connectors/hive/FileHandle.h [hunk 6]
```diff
@@ -67,14 +67,35 @@ using FileHandleCache = SimpleLRUCache<std::string, FileHandle>;
 
 // Creates FileHandles via the Generator interface the CachedFactory requires.
 class FileHandleGenerator {
+ public:
+  struct Stats {
+    uint64_t minGenerationTimeMcrs;
+    uint64_t maxGenerationTimeMcrs;
+    uint64_t totalGenerationTimeMcrs;
+    size_t numGeneration;
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/connectors/hive/FileHandle.h]: will be gone

File: velox/connectors/hive/FileHandle.h [hunk 7]
```diff
@@ -67,14 +67,35 @@ using FileHandleCache = SimpleLRUCache<std::string, FileHandle>;
...
+        numGeneration_};
+  }
 
  private:
   const std::shared_ptr<const Config> properties_;
+  std::mutex m_;
+  uint64_t minGenerationTimeMcrs_{std::numeric_limits<uint64_t>::max()};
+  uint64_t maxGenerationTimeMcrs_{0};
+  uint64_t totalGenerationTimeMcrs_{0};
+  size_t numGeneration_{0};
```
[the remark this answers is not in the document]
Comment (tanjialiang) [reply velox/connectors/hive/FileHandle.h]: Yup will be gone, too

File: velox/connectors/hive/FileHandle.h (hunk 1 above)
Comment (tanjialiang) [reply velox/connectors/hive/FileHandle.h]: will be gone, too

File: velox/common/base/Counters.h [hunk 8]
```diff
@@ -15,9 +15,14 @@
...
+
+#include <folly/Range.h>
+
 namespace facebook::velox {
 
 // Velox Counter Registration
-void registerVeloxCounters() {}
+void registerVeloxCounters();
 
+constexpr folly::StringPiece kCounterFileHandleGenerateLatencyMs{
```
Comment (xiaoxmeng) [inline velox/common/base/Counters.h:26]:
```
s/kCounterFileHandleGenerateLatencyMs/kCounterHiveFileHandleGenerateLatencyMs/
s/velox.file_handle_generate_latency_ms/velox.hive.file_handle_generate_latency_ms/
```

File: velox/common/base/Counters.cpp [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+#include "velox/common/base/Counters.h"
+#include "velox/common/base/StatsReporter.h"
+
+namespace facebook::velox {
+
+void registerVeloxCounters() {
+  REPORT_ADD_HISTOGRAM_EXPORT_PERCENTILE(
```
Comment (xiaoxmeng) [inline velox/common/base/Counters.cpp:23]:
```
// Track hive handle generation latency in range of [0, 100s] and reports P50, P90, P95 and P99.
kCounterHiveFileHandleGenerateLatencyMs, 10, 0, 100'000, 50, 90, 95, 99);
```

File: velox/common/caching/SimpleLRUCache.h [hunk 10]
```diff
@@ -36,125 +26,147 @@ namespace facebook::velox {
...
 
   // Current cache size used by pinned entries.
-  int64_t pinnedSize{0};
+  size_t pinnedSize{0};
 
   // Total number of elements in the cache.
   size_t numElements{0};
 
+  // Total number of cache hits since server start.
+  size_t numAccumulativeHits{0};
```
Comment (xiaoxmeng) [inline velox/common/caching/SimpleLRUCache.h:41]:
```
nit: s/numAccumulativeHits/numHits/
```

File: velox/common/caching/SimpleLRUCache.h [hunk 11]
```diff
@@ -36,125 +26,147 @@ namespace facebook::velox {
...
+
+  // Total number of cache fetches since server start, either hit or not.
+  size_t numAccumulativeRequests{0};
```
Comment (xiaoxmeng) [inline velox/common/caching/SimpleLRUCache.h:44]:
```
// Total number of cache lookups ...

s/numAccumulativeRequests/numRequests or numLookups/
Lookup is common term used for cache.
```

File: velox/common/caching/SimpleLRUCache.h [hunk 12]
```diff
@@ -36,125 +26,147 @@ namespace facebook::velox {
...
-    stats.maxSize = maxSize_;
-    stats.curSize = curSize_;
-    stats.pinnedSize = pinnedSize_;
-    return stats;
+    return {
+        maxSize_,
+        curSize_,
+        pinnedSize_,
+        elements_.size(),
+        numAccumulativeHits_,
```
Comment (xiaoxmeng) [inline velox/common/caching/SimpleLRUCache.h:146]:
```
s/numAccumulativeHits_/numHits_/
s/numAccumulativeRequests_/numLookups_/
```

File: velox/connectors/hive/FileHandle.cpp [hunk 13]
```diff
@@ -38,13 +41,20 @@ std::string groupName(const std::string& filename) {
...
 std::unique_ptr<FileHandle> FileHandleGenerator::operator()(
     const std::string& filename) {
-  auto fileHandle = std::make_unique<FileHandle>();
-  fileHandle->file = filesystems::getFileSystem(filename, properties_)
-                         ->openFileForRead(filename);
-  fileHandle->uuid = StringIdLease(fileIds(), filename);
-  fileHandle->groupId = StringIdLease(fileIds(), groupName(filename));
-  VLOG(1) << "Generating file handle for: " << filename
-          << " uuid: " << fileHandle->uuid.id();
+  uint64_t elapsedTimeMcrs{0};
```
Comment (xiaoxmeng) [inline velox/connectors/hive/FileHandle.cpp:44]:
```
s/elapsedTimeMcrs/elapsedTimeUs/
```

**Review by xiaoxmeng — APPROVED** [review]
@tanjialiang thanks for the update. LGTM!

**Pull request closed without merging**